### DISCUSSION TITLE
pandas dataframe table rendering: how to remove row number?

### DISCUSSION THREAD

**Discussion by aborruso** [quarto-dev/quarto-cli#3456] [Q&A] [answered]
Hi,
I have something similar to the below image.

It's the simply output of

```python
df= pd.read_csv("input.csv")
df
```

Is it a quarto way to style the table and in example hide/remove the  row index, that it's not real data.

Thank you

![image](https://user-images.githubusercontent.com/30607/204005803-5a4ec680-0b42-4c9f-9f7b-f019723bc08e.png)

**Comment by cscheid** [comment]
This is a pandas index rendering issue. https://stackoverflow.com/questions/24644656/how-to-print-pandas-dataframe-without-index

**Reply by aborruso** [reply]
> This is a pandas index rendering issue. https://stackoverflow.com/questions/24644656/how-to-print-pandas-dataframe-without-index

Thank you @cscheid I had already tried it.

But in the second way, I have to find a way to get the quarto pandas dataframe standard style, and not to have a sort of plain text.

![image](https://user-images.githubusercontent.com/30607/204145268-e1530f79-476a-4c92-a8a8-07838e3bfeb0.png)

**Comment by AlFontal** [comment, marked as answer]
As @cscheid mentioned, this is something related to `pandas` and not `quarto`.

To solve this (and much more) in `pandas`, check the [style guide](https://pandas.pydata.org/docs/user_guide/style.html), which enables you to customize almost everything in a rendered dataframe.

For your case, what you should do is simply:

```python

df.style.hide(axis='index')

```

That should remove the index for the output and give you the default rendering.

**Reply by aborruso** [reply]
> As @cscheid mentioned, this is something related to `pandas` and not `quarto`.

First of all thank you @AlFontal.

I understand what you and @cscheid  are saying, I don't want to waste your time.

But I insist, because in my opinion, as a user it's also something that concerns quarto.
Try to give one last example and then I'll stop, because I respect you very much and I don't want to disturb you.

Ok, I have removed the row index using `df.style.hide_index()`.

But I have two points:
- quarto is a really great and impressive scientific and technical publishing system. It should have a base method (one for each language), in which to read an external table and print it. The pandas row index is not in the data;
- if I remove the index, I do not have no more the standard pandas striped markdown table, I have another style, without alternate colors, without fit to width (the first one).

Having said that, I'll stop here, thank you very much and I must learn to use the styles

![image](https://user-images.githubusercontent.com/30607/204149127-acb523e0-ce0b-44a5-8fdc-cee90a016652.png)

**Reply by AlFontal** [reply]
Oh, I see. As far as I understand it:

+ I disagree with you about the fact that `quarto` should be aware of the 'index' of a table. It would be impossible to accomodate every library and language and would overcomplicate the issue. As far as `quarto` or any other framework is aware, the `index` is simply an extra column (or columns, in case of a `MultiIndex`). It _actually_ belongs to the data. If you want a really hacky solution you could always set one of your columns as your index, but I know it's not what you want.

+ I agree with you in that the fact that `quarto` is unable to parse and generate a nice render of `pandas` styled outputs _is_ a problem. I think the output of `df.style.hide(axis='index')` should provide a nicely rendered table. I was a heavy user of `jupyterbook` and it integrated perfectly (you can do a lot with the `style` methods, not only hiding the indexes).

**Reply by AlFontal** [reply]
Hey @aborruso, I actually found out how to solve this a while ago, I am now having some _related_ problem with how quarto handles CSS generated by the `pandas` styler.

That being said, the representation of the dataframe goes wrong because the output of the `style` method renders html without the `dataframe` subclass. You can however easily fix this by appending an extra method:

```python
df.style.hide_index().set_table_attributes("class='dataframe'")
```

In fact, you can add any extra classes you wish here, for instance here you add the bootstrap  'hover' effect that many tables usually have (and center the text, if you wish):

```python
 (df
 .style
 .hide_index()
 .set_table_attributes("class='dataframe table-hover', style='text-align: center'")
)
```

**Reply by aborruso** [reply]
Thank you @AlFontal

**Reply by AlFontal** [reply]
No problem @aborruso 

I just opened my discussion on the related issue #3964 

If you mark my reply as 'answered' it will give me some nice virtual worthless points but will close the discussion hahaha

**Reply by aborruso** [reply]
> If you mark my reply as 'answered' it will give me some nice virtual worthless points but will close the discussion hahaha

Sure!!!

**Reply by aborruso** [reply]
@AlFontal I need to understand one more thing.

If I use `set_table_styles`

~~~~
```{python}
#| echo: false
#| warning: false

import pandas as pd

df = pd.DataFrame({'text': ['foo', 'bar'],'number': [1, 2]})

dfStyler = df.style.set_properties(**{'text-align': 'center'})
dfStyler.set_table_styles([dict(selector='th', props=[('text-align', 'center')])])
```
~~~~

I have in the notebook the style applied

![image](https://user-images.githubusercontent.com/30607/213640152-cd282f9e-d7cd-4cbd-aa52-6b40188ee199.png)

If I run it in quarto, styling is managed by bootstrap, and I have the default left align

![image](https://user-images.githubusercontent.com/30607/213640307-f60b92c6-ad34-4028-b884-7dfb1f99dc33.png)

Is it useless then to use set_table_styles here?

I can apply in example

```
df.style.hide_index().set_table_attributes("class='dataframe table-hover text-center'")
```

to have 

![image](https://user-images.githubusercontent.com/30607/213642910-9d2c13da-8aa2-4b94-87c9-d960e4ba8c27.png)


But how to apply `text-center` only to `td`, or moreover how to align left the `text` column and to the right the `number` column?

In example I cannot use

```
df.style.set_properties(subset=["number"], **{'text-align': 'right'})
```

It works in the notebook, but in quarto bootstrap forces number` column to align to the left

![image](https://user-images.githubusercontent.com/30607/213643738-7eff44e2-e4d6-4b1c-b7f7-145a57438c8b.png)

Thank you

**Reply by AlFontal** [reply]
Hey @aborruso!

The way Quarto handles the CSS output of `pandas` styler is a bit annoying yes, my discussion (#3964 ) got no traction whatsoever sadly. 

The thing is that the pandas styler appends a <style> tag with CSS specifying the attributes for the elements within the table, but Quarto, when rendering from `.ipynb` to `.html` completely omits that tag.

There are two ways of fixing them, one is a bit cumbersome, and the other is even more cumbersome:

+ If you want a table class where the `td` are centered and the `th` are left-aligned, you can just define custom CSS classes and include the file in your `_quarto.yml`:

```yml
# _quarto.yml
format:
  html:
    css: styles.css
```

```css

/* styles.css */

td td-left {
  text-align: left;
}
th .th-center {
  text-align: center;
}
```

And then, include these defined classes in your call:

```python
(df
 .style
 .hide_index()
 .set_table_attributes("class='dataframe table-hover td-left th-center'")
)
```

This seems to work on my examples. But you will not be able to select custom formatting for each different column, you can only differentiate between headers and normal cells. 

If you want to make use of the extensive CSS editing that `pandas` provides and customize individual cells:

Here I am doing a really cumbersome and ugly solution, but it's working for my usecase. Basically, you have to save the custom styles generated by pandas into a CSS file and include them in the `.yml` metadata of your document. 

In my case, I use BeautifulSoup to extract the style tag from the generated html since it's already a dependency of my project, but you should be able to do the same manually. The example I generated in my discussion 'solved', showing you an example where each :

```python

from bs4 import BeautifulSoup
from plotnine.data import mtcars

df = (mtcars
[['name', 'mpg', 'cyl']]
 .head()
 .style
 .hide(axis='index')
 .set_table_attributes("class='dataframe'")
 .highlight_max(color='red')
 .set_properties(subset=['name'], **{'text-align': 'left'})
 .set_properties(subset=['mpg'], **{'text-align': 'right'})
 .set_properties(subset=['cyl'], **{'text-align': 'center'})
 .format({'mpg': '{:.2f}'})
)

style = BeautifulSoup(df.to_html()).find('style')
with open('custom.css', 'w') as fh:
    fh.write(style.text)
df

```
This renders in Quarto:

<img width="316" alt="imagen" src="https://user-images.githubusercontent.com/10958332/213684904-10bd2f4b-490b-4f12-a611-83c9f69bdff4.png">


Basically what I do is keep a file per document and append to the `custom.css` stylesheet everytime I use any extra CSS.

It's really cumbersome so if it was fixed by default when quarto renders it it would be great, but at least this seems to work. If you can upvote or comment on my discussion (#3964) maybe we can get some attention from the devs.

**Reply by aborruso** [reply]
The sad thing is that the table formatting you apply in R, it is easy and immediate to have it in the quarto html output. Not with python :(

@cscheid I'm tagging you, because you are in this thread, and I think that #3964 is useful and important.

Thank you very much @AlFontal

**Reply by aborruso** [reply]
@AlFontal I have tried to apply your great tip

```
(df.style.hide_index().set_table_attributes("class='dataframe table-hover'"))
```

to apply `table-hover` to a table. But it does not work, I do not have `table-hover` style.
What am I doing wrong?

I'm using quarto 1.3.228. This is [the rendered HTML page](https://pnrr.datibenecomune.it/post/risorse-pnrr-regione-piemonte/), this is [the code](https://github.com/ondata/italian-public-sector-pnrr-data-guide/blob/main/post/risorse-pnrr-regione-piemonte/index.qmd#L23-L34) and this the [_metadata](https://github.com/ondata/italian-public-sector-pnrr-data-guide/blob/main/post/_metadata.yml) for that folder.

As usual, thank you

**Reply by aborruso** [reply]
@AlFontal it was my stupid error I had `#| output: false`. Sorry

**Reply by AlFontal** [reply]
Glad you got it solved @aborruso. That's a nice website!
Buon inizio di settimana ;)
